x-pack/plugin/compat-rest-request/src/main/java/org/elasticsearch/compat/CompatRestRequest.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.compat;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.plugins.RestCompatibility;
+
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CompatRestRequest extends Plugin implements RestCompatibility {
+
+    private static final Pattern COMPATIBLE_API_HEADER_PATTERN = Pattern.compile(
+        "(application|text)/(vnd.elasticsearch\\+)?([^;]+)(\\s*;\\s*compatible-with=(\\d+))?",
+        Pattern.CASE_INSENSITIVE);
+
+    @Override
+    public Version getCompatibleVersion(@Nullable String acceptHeader, @Nullable String contentTypeHeader) {
```

Comment:
it does look better than the logic we had before. Lack of hasContent() makes it look cleaner, but makes the validation less strict. I will add previous test cases to verify